fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/BaseScalarOperatorShuttle.java

```diff
@@ -0,0 +1,216 @@
+        } else {
+            return predicate;
+        }
+    }
+
+    @Override
+    public ScalarOperator visitInPredicate(InPredicateOperator predicate, Void context) {
+        boolean[] update = {false};
+        List<ScalarOperator> clonedOperators = visitList(predicate.getChildren(), update);
+        if (update[0]) {
+            return new InPredicateOperator(predicate.isNotIn(), clonedOperators);
+        } else {
+            return predicate;
+        }
+    }
+
+    @Override
+    public ScalarOperator visitIsNullPredicate(IsNullPredicateOperator predicate, Void context) {
+        boolean[] update = {false};
+        List<ScalarOperator> clonedOperators = visitList(predicate.getChildren(), update);
+        if (update[0]) {
+            return new IsNullPredicateOperator(predicate.isNotNull(), clonedOperators.get(0));
+        } else {
+            return predicate;
+        }
+    }
+
+    @Override
+    public ScalarOperator visitLikePredicateOperator(LikePredicateOperator predicate, Void context) {
+        boolean[] update = {false};
+        List<ScalarOperator> clonedOperators = visitList(predicate.getChildren(), update);
+        if (update[0]) {
+            return new LikePredicateOperator(predicate.getLikeType(), clonedOperators);
+        } else {
+            return predicate;
+        }
+    }
+
+    @Override
+    public ScalarOperator visitCastOperator(CastOperator operator, Void context) {
+        boolean[] update = {false};
+        List<ScalarOperator> clonedOperators = visitList(operator.getChildren(), update);
+        if (update[0]) {
+            return new CastOperator(operator.getType(), clonedOperators.get(0), operator.isImplicit());
+        } else {
+            return operator;
+        }
+    }
+
+    @Override
+    public ScalarOperator visitCaseWhenOperator(CaseWhenOperator operator, Void context) {
+        boolean[] update = {false};
+        List<ScalarOperator> clonedOperators = visitList(operator.getChildren(), update);
+        if (update[0]) {
+            ScalarOperator clonedCaseClause = null;
+            ScalarOperator clonedElseClause = null;
+            List<ScalarOperator> clonedWhenThenClauses;
+            if (operator.getCaseClause() != null) {
+                clonedCaseClause = clonedOperators.get(0);
+            }
+            if (operator.getElseClause() != null) {
+                clonedElseClause = clonedOperators.get(clonedOperators.size() - 1);
+            }
+
+            int whenThenEndIdx = operator.getElseClause() == null ? clonedOperators.size() : clonedOperators.size() - 1;
+            clonedWhenThenClauses = clonedOperators.subList(operator.getWhenStart(), whenThenEndIdx);
+
+            return new CaseWhenOperator(operator.getType(), clonedCaseClause, clonedElseClause, clonedWhenThenClauses);
+        } else {
+            return operator;
+        }
+    }
+
+    protected List<ScalarOperator> visitList(List<ScalarOperator> operators, boolean[] update) {
```

Comment:
seems you only use  update[0], why we need a boolean[] here?

Reply:
I need visitList return the update flag info at the same time. boolean is primitive data type, so I use a array to pass the reference.

Reply:
so, why not Boolean?